File: tests/support/fixture.rs
use std::fmt;
use std::fs;
use std::panic::{catch_unwind, AssertUnwindSafe};
use std::path::Path;

use hugr_lean::engine::Engine;
use hugr_lean::normalize::terminal_safe_text;
use hugr_lean::protocol::{
    CompletenessV1, DecisionV1, ObservationV1, PresentationV1, ShellDialectV1, SourceV1,
    TerminationKindV1, TerminationV1, PROTOCOL_V1,
};
use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FixtureCase {
    pub schema: u16,
    pub id: String,
    pub kind: FixtureKind,
    pub observation: FixtureObservation,
    pub expect: FixtureExpectation,
    pub preservation: FixturePreservation,
    pub provenance: FixtureProvenance,
    pub normalization: Option<FixtureNormalization>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FixtureNormalization {
    pub primitive: NormalizationPrimitive,
}

#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum NormalizationPrimitive {
    StripSgr,
    CollapseCarriageRedraws,
}

#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FixtureKind {
    Core,
    Normalization,
    Profile,
    Integration,
    Regression,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FixtureObservation {
    pub source: SourceV1,
    pub command: String,
    pub shell_dialect: FixtureShellDialect,
    pub termination: TerminationKindV1,
    pub exit_code: Option<i32>,
    pub completeness: CompletenessV1,
    pub presentation: PresentationV1,
}

#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FixtureShellDialect {
    Unknown,
    Posix,
    Powershell,
    Cmd,
}

impl From<FixtureShellDialect> for ShellDialectV1 {
    fn from(value: FixtureShellDialect) -> Self {
        match value {
            FixtureShellDialect::Unknown => Self::Unknown,
            FixtureShellDialect::Posix => Self::Posix,
            FixtureShellDialect::Powershell => Self::PowerShell,
            FixtureShellDialect::Cmd => Self::Cmd,
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FixtureExpectation {
    pub decision: DecisionV1,
    pub profile: String,
    pub golden: Option<String>,
    #[serde(default)]
    pub required_literals: Vec<String>,
    #[serde(default)]
    pub forbidden_literals: Vec<String>,
    #[serde(default)]
    pub properties: Vec<FixtureProperty>,
}

#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FixtureProperty {
    NonExpanding,
    Idempotent,
    PassthroughExact,
    NoPanic,
    PreservesRequiredLiterals,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FixturePreservation {
    #[serde(default)]
    pub mandatory_signal_ids: Vec<String>,
    #[serde(default)]
    pub permitted_removals: Vec<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct FixtureProvenance {
    pub kind: FixtureProvenanceKind,
    pub source_repo: String,
    pub source_commit: String,
    pub source_path: String,
    pub license: String,
    pub origin_issue: u64,
    pub notes: String,
}

#[derive(Debug, Clone, Copy, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FixtureProvenanceKind {
    Synthetic,
    Captured,
    Donor,
    Regression,
}

#[derive(Debug, Clone)]
pub struct LoadedFixture {
    pub case: FixtureCase,
    pub input: String,
    pub expected: Option<String>,
}

impl LoadedFixture {
    pub fn load(root: impl AsRef<Path>) -> Result<Self, HarnessError> {
        let root = root.as_ref().to_path_buf();
        let case_text = read(&root.join("case.toml"))?;
        let case = parse_case_toml(&case_text)?;

        if case.schema != 1 {
            return Err(HarnessError::new(format!(
                "fixture {} uses unsupported schema {}",
                case.id, case.schema
            )));
        }
        if case.id.is_empty() {
            return Err(HarnessError::new("fixture id must not be empty"));
        }

        let directory_id = root
            .file_name()
            .and_then(|value| value.to_str())
            .ok_or_else(|| HarnessError::new("fixture directory must have a UTF-8 name"))?;
        if directory_id != case.id {
            return Err(HarnessError::new(format!(
                "fixture id {:?} does not match directory {:?}",
                case.id, directory_id
            )));
        }

        validate_kind_and_provenance(&case)?;
        validate_normalization_metadata(&case)?;
        validate_provenance(&case)?;
        validate_preservation_metadata(&case)?;

        let input = read(&root.join("input.txt"))?;
        let expected = match &case.expect.golden {
            Some(path) => {
                validate_relative_fixture_path(path)?;
                Some(read(&root.join(path))?)
            }
            None => None,
        };

        Ok(Self {
            case,
            input,
            expected,
        })
    }

    pub fn observation(&self) -> Result<ObservationV1, HarnessError> {
        let termination = TerminationV1 {
            kind: self.case.observation.termination,
            code: self.case.observation.exit_code,
        };
        termination
            .validate()
            .map_err(|error| HarnessError::new(error.to_string()))?;

        let observation = ObservationV1 {
            schema_version: PROTOCOL_V1,
            source: self.case.observation.source,
            command: nonempty(&self.case.observation.command),
            shell_dialect: self.case.observation.shell_dialect.into(),
            output: self.input.clone(),
            termination,
            completeness: self.case.observation.completeness,
            presentation: self.case.observation.presentation,
        };
        observation
            .validate()
            .map_err(|error| HarnessError::new(error.to_string()))?;
        Ok(observation)
    }
}

pub fn parse_case_toml(input: &str) -> Result<FixtureCase, HarnessError> {
    toml::from_str(input)
        .map_err(|error| HarnessError::new(format!("invalid fixture TOML: {error}")))
}

pub fn verify_fixture(engine: &Engine, fixture: &LoadedFixture) -> Result<(), HarnessError> {
    let observation = fixture.observation()?;
    let input = observation.output.clone();

    let result = catch_unwind(AssertUnwindSafe(|| engine.process(observation.clone())))
        .map_err(|_| HarnessError::new(format!("fixture {} panicked", fixture.case.id)))?
        .map_err(|error| HarnessError::new(error.to_string()))?;

    if result.decision != fixture.case.expect.decision {
        return Err(HarnessError::new(format!(
            "fixture {} expected decision {:?}, got {:?}",
            fixture.case.id, fixture.case.expect.decision, result.decision
        )));
    }

    let expected_profile = nonempty(&fixture.case.expect.profile);
    if result.profile.as_deref() != expected_profile.as_deref() {
        return Err(HarnessError::new(format!(
            "fixture {} expected profile {:?}, got {:?}",
            fixture.case.id, expected_profile, result.profile
        )));
    }

    let effective = result.replacement.as_deref().unwrap_or(&input);

    if let Some(expected) = &fixture.expected {
        if effective != expected {
            return Err(HarnessError::new(format!(
                "fixture {} golden output mismatch",
                fixture.case.id
            )));
        }
    }

    for literal in &fixture.case.expect.required_literals {
        if !effective.contains(literal) {
            return Err(HarnessError::new(format!(
                "fixture {} lost required literal {:?}",
                fixture.case.id, literal
            )));
        }
    }

    for literal in &fixture.case.expect.forbidden_literals {
        if effective.contains(literal) {
            return Err(HarnessError::new(format!(
                "fixture {} retained forbidden literal {:?}",
                fixture.case.id, literal
            )));
        }
    }

    for property in &fixture.case.expect.properties {
        verify_property(*property, engine, &observation, &result, effective)?;
    }

    Ok(())
}

pub fn verify_normalization_fixture(fixture: &LoadedFixture) -> Result<(), HarnessError> {
    if fixture.case.kind != FixtureKind::Normalization {
        return Err(HarnessError::new(format!(
            "fixture {} is not a normalization fixture",
            fixture.case.id
        )));
    }

    let normalization = fixture
        .case
        .normalization
        .as_ref()
        .ok_or_else(|| HarnessError::new("normalization fixture missing primitive"))?;

    let observation = fixture.observation()?;
    let input = observation.output.clone();

    let effective = catch_unwind(AssertUnwindSafe(|| {
        let Some(terminal) = terminal_safe_text(&observation) else {
            return input.clone();
        };

        match normalization.primitive {
            NormalizationPrimitive::StripSgr => terminal.strip_sgr(),
            NormalizationPrimitive::CollapseCarriageRedraws => {
                terminal.collapse_carriage_redraws()
            }
        }
    }))
    .map_err(|_| HarnessError::new(format!("fixture {} panicked", fixture.case.id)))?;

    let decision = if effective == input {
        DecisionV1::Passthrough
    } else {
        DecisionV1::Normalized
    };

    if decision != fixture.case.expect.decision {
        return Err(HarnessError::new(format!(
            "fixture {} expected decision {:?}, got {:?}",
            fixture.case.id, fixture.case.expect.decision, decision
        )));
    }

    if !fixture.case.expect.profile.is_empty() {
        return Err(HarnessError::new(
            "normalization fixtures must not declare a profile",
        ));
    }

    if let Some(expected) = &fixture.expected {
        if &effective != expected {
            return Err(HarnessError::new(format!(
                "fixture {} golden output mismatch",
                fixture.case.id
            )));
        }
    }

    for literal in &fixture.case.expect.required_literals {
        if !effective.contains(literal) {
            return Err(HarnessError::new(format!(
                "fixture {} lost required literal {:?}",
                fixture.case.id, literal
            )));
        }
    }

    for literal in &fixture.case.expect.forbidden_literals {
        if effective.contains(literal) {
            return Err(HarnessError::new(format!(
                "fixture {} retained forbidden literal {:?}",
                fixture.case.id, literal
            )));
        }
    }

    for property in &fixture.case.expect.properties {
        match property {
            FixtureProperty::NonExpanding => {
                if effective.len() > input.len() {
                    return Err(HarnessError::new("non_expanding property failed"));
                }
            }
            FixtureProperty::Idempotent => {
                let mut second_observation = observation.clone();
                second_observation.output = effective.clone();
                let second = match terminal_safe_text(&second_observation) {
                    Some(terminal) => match normalization.primitive {
                        NormalizationPrimitive::StripSgr => terminal.strip_sgr(),
                        NormalizationPrimitive::CollapseCarriageRedraws => {
                            terminal.collapse_carriage_redraws()
                        }
                    },
                    None => effective.clone(),
                };
                if second != effective {
                    return Err(HarnessError::new("idempotence property failed"));
                }
            }
            FixtureProperty::PassthroughExact => {
                if decision != DecisionV1::Passthrough || effective != input {
                    return Err(HarnessError::new("passthrough_exact property failed"));
                }
            }
            FixtureProperty::NoPanic => {}
            FixtureProperty::PreservesRequiredLiterals => {}
        }
    }

    Ok(())
}

fn verify_property(
    property: FixtureProperty,
    engine: &Engine,
    observation: &ObservationV1,
    result: &hugr_lean::protocol::FilterResultV1,
    effective: &str,
) -> Result<(), HarnessError> {
    match property {
        FixtureProperty::NonExpanding => {
            if effective.len() > observation.output.len()
                || result.metrics.output_bytes > result.metrics.input_bytes
            {
                return Err(HarnessError::new("non_expanding property failed"));
            }
        }
        FixtureProperty::Idempotent => {
            let mut second_observation = observation.clone();
            second_observation.output = effective.to_owned();
            let second = catch_unwind(AssertUnwindSafe(|| engine.process(second_observation)))
                .map_err(|_| HarnessError::new("idempotence re-run panicked"))?
                .map_err(|error| HarnessError::new(error.to_string()))?;
            let second_effective = second.replacement.as_deref().unwrap_or(effective);
            if second_effective != effective {
                return Err(HarnessError::new("idempotence property failed"));
            }
        }
        FixtureProperty::PassthroughExact => {
            if result.decision != DecisionV1::Passthrough || effective != observation.output {
                return Err(HarnessError::new("passthrough_exact property failed"));
            }
        }
        FixtureProperty::NoPanic => {
            // The main verification call is already wrapped in catch_unwind.
        }
        FixtureProperty::PreservesRequiredLiterals => {
            // Required literals are checked unconditionally above.
        }
    }

    Ok(())
}

fn validate_normalization_metadata(case: &FixtureCase) -> Result<(), HarnessError> {
    match (case.kind, case.normalization.as_ref()) {
        (FixtureKind::Normalization, Some(_)) => Ok(()),
        (FixtureKind::Normalization, None) => Err(HarnessError::new(
            "normalization fixture kind requires [normalization] metadata",
        )),
        (_, Some(_)) => Err(HarnessError::new(
            "[normalization] metadata requires fixture kind = normalization",
        )),
        (_, None) => Ok(()),
    }
}

fn validate_kind_and_provenance(case: &FixtureCase) -> Result<(), HarnessError> {
    match (case.kind, case.provenance.kind) {
        (FixtureKind::Regression, FixtureProvenanceKind::Regression) => Ok(()),
        (FixtureKind::Regression, _) => Err(HarnessError::new(
            "regression fixture kind requires regression provenance",
        )),
        (_, FixtureProvenanceKind::Regression) => Err(HarnessError::new(
            "regression provenance requires fixture kind = regression",
        )),
        _ => Ok(()),
    }
}

fn validate_preservation_metadata(case: &FixtureCase) -> Result<(), HarnessError> {
    if case
        .preservation
        .mandatory_signal_ids
        .iter()
        .any(|value| value.is_empty())
    {
        return Err(HarnessError::new(
            "mandatory_signal_ids must not contain empty values",
        ));
    }

    if case
        .preservation
        .permitted_removals
        .iter()
        .any(|value| value.is_empty())
    {
        return Err(HarnessError::new(
            "permitted_removals must not contain empty values",
        ));
    }

    Ok(())
}

fn validate_relative_fixture_path(value: &str) -> Result<(), HarnessError> {
    let path = Path::new(value);
    if path.is_absolute()
        || path
            .components()
            .any(|component| matches!(component, std::path::Component::ParentDir))
    {
        return Err(HarnessError::new(
            "fixture golden path must remain inside the fixture directory",
        ));
    }

    Ok(())
}

fn validate_provenance(case: &FixtureCase) -> Result<(), HarnessError> {
    match case.provenance.kind {
        FixtureProvenanceKind::Synthetic => Ok(()),
        FixtureProvenanceKind::Captured => {
            if case.provenance.notes.is_empty() {
                Err(HarnessError::new(
                    "captured fixture provenance requires reproduction notes",
                ))
            } else {
                Ok(())
            }
        }
        FixtureProvenanceKind::Donor => {
            if [
                &case.provenance.source_repo,
                &case.provenance.source_commit,
                &case.provenance.source_path,
                &case.provenance.license,
            ]
            .iter()
            .any(|value| value.is_empty())
            {
                Err(HarnessError::new(
                    "donor fixture provenance requires repo, commit, path, and license",
                ))
            } else {
                Ok(())
            }
        }
        FixtureProvenanceKind::Regression => {
            if case.provenance.origin_issue == 0 {
                Err(HarnessError::new(
                    "regression fixture provenance requires origin_issue",
                ))
            } else {
                Ok(())
            }
        }
    }
}

fn read(path: &Path) -> Result<String, HarnessError> {
    fs::read_to_string(path)
        .map_err(|error| HarnessError::new(format!("{}: {error}", path.display())))
}

fn nonempty(value: &str) -> Option<String> {
    if value.is_empty() {
        None
    } else {
        Some(value.to_owned())
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HarnessError(String);

impl HarnessError {
    fn new(message: impl Into<String>) -> Self {
        Self(message.into())
    }
}

impl fmt::Display for HarnessError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl std::error::Error for HarnessError {}

File: tests/support/fixture.ts
import fs from "node:fs";
import path from "node:path";

import { parse } from "smol-toml";

import { Engine } from "../../src/engine.js";
import {
  collapseMonotonicAsciiRedraws,
  stripRecognizedSgr,
} from "../../src/normalize.js";
import {
  PROTOCOL_V1,
  exited,
  unknownTermination,
  utf8Bytes,
  validateObservationV1,
  type DecisionV1,
  type ObservationV1,
  type PresentationV1,
  type ShellDialectV1,
  type SourceV1,
} from "../../src/types.js";

export type FixtureKind =
  | "core"
  | "normalization"
  | "profile"
  | "integration"
  | "regression";

export type FixtureProperty =
  | "non_expanding"
  | "idempotent"
  | "passthrough_exact"
  | "no_panic"
  | "preserves_required_literals";

export type NormalizationPrimitive =
  | "strip_sgr"
  | "collapse_carriage_redraws";

export interface FixtureCase {
  schema: number;
  id: string;
  kind: FixtureKind;
  observation: {
    source: SourceV1;
    command: string;
    shell_dialect: "unknown" | "posix" | "powershell" | "cmd";
    termination: "unknown" | "exited" | "aborted" | "timed_out";
    exit_code?: number;
    completeness: ObservationV1["completeness"];
    presentation: PresentationV1;
  };
  expect: {
    decision: DecisionV1;
    profile: string;
    golden?: string;
    required_literals: string[];
    forbidden_literals: string[];
    properties: FixtureProperty[];
  };
  preservation: {
    mandatory_signal_ids: string[];
    permitted_removals: string[];
  };
  provenance: {
    kind: "synthetic" | "captured" | "donor" | "regression";
    source_repo: string;
    source_commit: string;
    source_path: string;
    license: string;
    origin_issue: number;
    notes: string;
  };
  normalization?: {
    primitive: NormalizationPrimitive;
  };
}

export interface LoadedFixture {
  root: string;
  case: FixtureCase;
  input: string;
  expected?: string;
}

export function loadFixture(root: string): LoadedFixture {
  const caseText = fs.readFileSync(path.join(root, "case.toml"), "utf8");
  const fixtureCase = parseCaseToml(caseText);
  validateFixtureCase(fixtureCase);

  if (path.basename(root) !== fixtureCase.id) {
    throw new Error(
      `fixture id ${fixtureCase.id} does not match directory ${path.basename(root)}`,
    );
  }

  const input = fs.readFileSync(path.join(root, "input.txt"), "utf8");
  let expected: string | undefined;
  if (fixtureCase.expect.golden !== undefined) {
    const golden = fixtureCase.expect.golden;
    validateRelativeFixturePath(golden);
    expected = fs.readFileSync(path.join(root, golden), "utf8");
  }

  return expected === undefined
    ? { root, case: fixtureCase, input }
    : { root, case: fixtureCase, input, expected };
}

export function fixtureObservation(fixture: LoadedFixture): ObservationV1 {
  const meta = fixture.case.observation;
  const termination =
    meta.termination === "exited"
      ? exited(requireExitCode(meta.exit_code))
      : {
          kind: meta.termination,
          code: meta.exit_code ?? null,
        };

  const observation: ObservationV1 = {
    schema_version: PROTOCOL_V1,
    source: meta.source,
    command: meta.command.length === 0 ? null : meta.command,
    shell_dialect: mapDialect(meta.shell_dialect),
    output: fixture.input,
    termination,
    completeness: meta.completeness,
    presentation: meta.presentation,
  };
  validateObservationV1(observation);
  return observation;
}

export function verifyFixture(engine: Engine, fixture: LoadedFixture): void {
  const observation = fixtureObservation(fixture);
  const result = engine.process(observation);

  if (result.decision !== fixture.case.expect.decision) {
    throw new Error(
      `${fixture.case.id}: expected ${fixture.case.expect.decision}, got ${result.decision}`,
    );
  }

  const expectedProfile =
    fixture.case.expect.profile.length === 0
      ? null
      : fixture.case.expect.profile;
  if (result.profile !== expectedProfile) {
    throw new Error(
      `${fixture.case.id}: expected profile ${String(expectedProfile)}, got ${String(result.profile)}`,
    );
  }

  const effective = result.replacement ?? fixture.input;
  verifyExpectedText(fixture, effective);

  for (const property of fixture.case.expect.properties) {
    if (property === "non_expanding") {
      if (
        utf8Bytes(effective) > utf8Bytes(fixture.input) ||
        result.metrics.output_bytes > result.metrics.input_bytes
      ) {
        throw new Error(`${fixture.case.id}: non_expanding failed`);
      }
    } else if (property === "idempotent") {
      const second = engine.process({ ...observation, output: effective });
      const secondEffective = second.replacement ?? effective;
      if (secondEffective !== effective) {
        throw new Error(`${fixture.case.id}: idempotence failed`);
      }
    } else if (property === "passthrough_exact") {
      if (result.decision !== "passthrough" || effective !== fixture.input) {
        throw new Error(`${fixture.case.id}: passthrough_exact failed`);
      }
    }
  }
}

export function verifyNormalizationFixture(fixture: LoadedFixture): void {
  if (fixture.case.kind !== "normalization" || fixture.case.normalization === undefined) {
    throw new Error("normalization metadata required");
  }

  const observation = fixtureObservation(fixture);
  const input = fixture.input;
  let effective = input;

  if (observation.presentation === "terminal_rendered") {
    effective =
      fixture.case.normalization.primitive === "strip_sgr"
        ? stripRecognizedSgr(input)
        : collapseMonotonicAsciiRedraws(input);
  }

  const decision: DecisionV1 =
    effective === input ? "passthrough" : "normalized";

  if (decision !== fixture.case.expect.decision) {
    throw new Error(
      `${fixture.case.id}: expected ${fixture.case.expect.decision}, got ${decision}`,
    );
  }

  if (fixture.case.expect.profile.length !== 0) {
    throw new Error("normalization fixture cannot declare profile");
  }

  verifyExpectedText(fixture, effective);

  for (const property of fixture.case.expect.properties) {
    if (property === "non_expanding" && utf8Bytes(effective) > utf8Bytes(input)) {
      throw new Error(`${fixture.case.id}: non_expanding failed`);
    }
    if (property === "idempotent") {
      const second =
        observation.presentation === "terminal_rendered"
          ? fixture.case.normalization.primitive === "strip_sgr"
            ? stripRecognizedSgr(effective)
            : collapseMonotonicAsciiRedraws(effective)
          : effective;
      if (second !== effective) {
        throw new Error(`${fixture.case.id}: idempotence failed`);
      }
    }
    if (
      property === "passthrough_exact" &&
      (decision !== "passthrough" || effective !== input)
    ) {
      throw new Error(`${fixture.case.id}: passthrough_exact failed`);
    }
  }
}

export function parseCaseToml(input: string): FixtureCase {
  return decodeFixtureCase(parse(input, { unsafeKeyBehaviour: "throw" }));
}

export function loadFixtureDirectories(parent: string): LoadedFixture[] {
  return fs
    .readdirSync(parent, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => loadFixture(path.join(parent, entry.name)))
    .sort((left, right) => left.case.id.localeCompare(right.case.id));
}

function decodeFixtureCase(value: unknown): FixtureCase {
  const root = record(value, "fixture");
  exactKeys(root, [
    "schema",
    "id",
    "kind",
    "observation",
    "expect",
    "preservation",
    "provenance",
    "normalization",
  ]);

  const observation = record(root.observation, "observation");
  exactKeys(observation, [
    "source",
    "command",
    "shell_dialect",
    "termination",
    "exit_code",
    "completeness",
    "presentation",
  ]);

  const expect = record(root.expect, "expect");
  exactKeys(expect, [
    "decision",
    "profile",
    "golden",
    "required_literals",
    "forbidden_literals",
    "properties",
  ]);

  const preservation = record(root.preservation, "preservation");
  exactKeys(preservation, ["mandatory_signal_ids", "permitted_removals"]);

  const provenance = record(root.provenance, "provenance");
  exactKeys(provenance, [
    "kind",
    "source_repo",
    "source_commit",
    "source_path",
    "license",
    "origin_issue",
    "notes",
  ]);

  const normalization =
    root.normalization === undefined
      ? undefined
      : decodeNormalization(root.normalization);

  const fixture: FixtureCase = {
    schema: integer(root.schema, "schema"),
    id: text(root.id, "id"),
    kind: enumeration(root.kind, ["core", "normalization", "profile", "integration", "regression"], "kind"),
    observation: {
      source: enumeration(observation.source, ["shell", "read", "search", "lsp", "mcp", "browser", "other"], "source"),
      command: text(observation.command, "command"),
      shell_dialect: enumeration(observation.shell_dialect, ["unknown", "posix", "powershell", "cmd"], "shell_dialect"),
      termination: enumeration(observation.termination, ["unknown", "exited", "aborted", "timed_out"], "termination"),
      completeness: enumeration(observation.completeness, ["unknown", "complete", "truncated"], "completeness"),
      presentation: enumeration(observation.presentation, ["unknown", "terminal_rendered"], "presentation"),
      ...(observation.exit_code === undefined
        ? {}
        : { exit_code: integer(observation.exit_code, "exit_code") }),
    },
    expect: {
      decision: enumeration(expect.decision, ["passthrough", "normalized", "reduced", "failed_open"], "decision"),
      profile: text(expect.profile, "profile"),
      ...(expect.golden === undefined ? {} : { golden: text(expect.golden, "golden") }),
      required_literals: stringArray(expect.required_literals, "required_literals"),
      forbidden_literals: stringArray(expect.forbidden_literals, "forbidden_literals"),
      properties: enumArray(
        expect.properties,
        ["non_expanding", "idempotent", "passthrough_exact", "no_panic", "preserves_required_literals"],
        "properties",
      ),
    },
    preservation: {
      mandatory_signal_ids: stringArray(preservation.mandatory_signal_ids, "mandatory_signal_ids"),
      permitted_removals: stringArray(preservation.permitted_removals, "permitted_removals"),
    },
    provenance: {
      kind: enumeration(provenance.kind, ["synthetic", "captured", "donor", "regression"], "provenance.kind"),
      source_repo: text(provenance.source_repo, "source_repo"),
      source_commit: text(provenance.source_commit, "source_commit"),
      source_path: text(provenance.source_path, "source_path"),
      license: text(provenance.license, "license"),
      origin_issue: integer(provenance.origin_issue, "origin_issue"),
      notes: text(provenance.notes, "notes"),
    },
    ...(normalization === undefined ? {} : { normalization }),
  };

  return fixture;
}

function decodeNormalization(value: unknown): { primitive: NormalizationPrimitive } {
  const normalization = record(value, "normalization");
  exactKeys(normalization, ["primitive"]);
  return {
    primitive: enumeration(
      normalization.primitive,
      ["strip_sgr", "collapse_carriage_redraws"],
      "normalization.primitive",
    ),
  };
}

function validateFixtureCase(fixture: FixtureCase): void {
  if (fixture.schema !== 1 || fixture.id.length === 0) {
    throw new Error("invalid fixture schema/id");
  }

  if (
    (fixture.kind === "normalization") !==
    (fixture.normalization !== undefined)
  ) {
    throw new Error("normalization metadata/kind mismatch");
  }

  if (
    (fixture.kind === "regression") !==
    (fixture.provenance.kind === "regression")
  ) {
    throw new Error("regression kind/provenance mismatch");
  }

  if (
    fixture.preservation.mandatory_signal_ids.some((value) => value.length === 0) ||
    fixture.preservation.permitted_removals.some((value) => value.length === 0)
  ) {
    throw new Error("empty preservation metadata");
  }

  if (fixture.provenance.kind === "captured" && fixture.provenance.notes.length === 0) {
    throw new Error("captured provenance needs notes");
  }
  if (
    fixture.provenance.kind === "donor" &&
    [
      fixture.provenance.source_repo,
      fixture.provenance.source_commit,
      fixture.provenance.source_path,
      fixture.provenance.license,
    ].some((value) => value.length === 0)
  ) {
    throw new Error("donor provenance incomplete");
  }
  if (
    fixture.provenance.kind === "regression" &&
    fixture.provenance.origin_issue === 0
  ) {
    throw new Error("regression provenance needs origin_issue");
  }
}

function verifyExpectedText(fixture: LoadedFixture, effective: string): void {
  if (fixture.expected !== undefined && effective !== fixture.expected) {
    throw new Error(`${fixture.case.id}: golden mismatch`);
  }
  for (const literal of fixture.case.expect.required_literals) {
    if (!effective.includes(literal)) {
      throw new Error(`${fixture.case.id}: lost required literal ${literal}`);
    }
  }
  for (const literal of fixture.case.expect.forbidden_literals) {
    if (effective.includes(literal)) {
      throw new Error(`${fixture.case.id}: retained forbidden literal ${literal}`);
    }
  }
}

function validateRelativeFixturePath(value: string): void {
  if (path.isAbsolute(value) || value.split(/[\\/]/u).includes("..")) {
    throw new Error("fixture golden path escapes fixture directory");
  }
}

function mapDialect(
  value: FixtureCase["observation"]["shell_dialect"],
): ShellDialectV1 {
  return value === "powershell" ? "power_shell" : value;
}

function requireExitCode(value: number | undefined): number {
  if (value === undefined) {
    throw new Error("exited termination requires exit_code");
  }
  return value;
}

function record(value: unknown, label: string): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error(`${label} must be a table`);
  }
  return value as Record<string, unknown>;
}

function exactKeys(
  value: Record<string, unknown>,
  allowed: readonly string[],
): void {
  const allowedSet = new Set(allowed);
  for (const key of Object.keys(value)) {
    if (!allowedSet.has(key)) {
      throw new Error(`unknown fixture field: ${key}`);
    }
  }
}

function text(value: unknown, label: string): string {
  if (typeof value !== "string") {
    throw new Error(`${label} must be string`);
  }
  return value;
}

function integer(value: unknown, label: string): number {
  if (typeof value !== "number" || !Number.isInteger(value)) {
    throw new Error(`${label} must be integer`);
  }
  return value;
}

function stringArray(value: unknown, label: string): string[] {
  if (!Array.isArray(value) || !value.every((item) => typeof item === "string")) {
    throw new Error(`${label} must be string array`);
  }
  return [...value] as string[];
}

function enumeration<const T extends readonly string[]>(
  value: unknown,
  allowed: T,
  label: string,
): T[number] {
  if (typeof value !== "string" || !allowed.includes(value)) {
    throw new Error(`${label} has invalid value`);
  }
  return value as T[number];
}

function enumArray<const T extends readonly string[]>(
  value: unknown,
  allowed: T,
  label: string,
): T[number][] {
  if (!Array.isArray(value)) {
    throw new Error(`${label} must be array`);
  }
  return value.map((item) => enumeration(item, allowed, label));
}
